### PR TITLE
Fix better_match check used by regex search routine

### PR DESCRIPTION
**Pull request by cpp11nullptr**
That is an attempt to fix the issue #731 by taking a number of captured groups into consideration during performing better_match check.

### CODE REVIEW

File: stl/inc/regex
Comment (StephanTLavavej) [inline stl/inc/regex:3431]: I have the following concerns:
1. Where is this mandated by ECMAScript? I am wary of adding logic to make this specific scenario work, without knowing exactly what Standardese it corresponds to.
2. Can we construct test cases that stress this rule (i.e. cases where it activates and causes different choices to be made), and compare the observed behavior of libc++, libstdc++, and browsers?
3. Does this affect the other grammars, and are they all specified to behave the same way?

In Boost's documentation (which is of course non-Standard, but more comprehensible than ECMAScript/POSIX/etc.) I see that the [Perl/ECMAScript behavior](https://www.boost.org/doc/libs/1_75_0/libs/regex/doc/html/boost_regex/syntax/perl_syntax.html#boost_regex.syntax.perl_syntax.what_gets_matched) is described as "the first match found by a depth-first-search" and "leftmost", while [basic/extended behavior](https://www.boost.org/doc/libs/1_75_0/libs/regex/doc/html/boost_regex/syntax/leftmost_longest_rule.html) is "leftmost longest".

**Comment by cpp11nullptr** [comment]
Thanks for additional input, @StephanTLavavej!

1. I could not find precise description what behaviour should be in such scenarios but this PR should correct a capturing for cases, such as "(A+)\s*(B+)?\s*B*", when for "AAA BBB" target there is firstly found "(A+)..B*" match with only (A+) group captured with writing  to _Res with no way to overwrite it even if "(A+)..(B+)" match (with two captured groups) found. Perhaps, it is because of implementation details and, as alternative, a logic could be changed to make "(A+)..(B+)" match found firstly (but it's not so robust).
2. Good point, I added few more examples:

```
#include <iostream>
#include <regex>
#include <string>

void find_match(const std::string& title, const std::string& rgstr, const std::string& str)
{
    std::cout << title << ":";

    const std::regex rg{ rgstr };
    std::smatch match{};
    if (!std::regex_match(str, match, rg))
    {
        std::cout << "No match." << std::endl;

        return;
    }

    for (size_t i{ 1 }; i < match.size(); ++i)
    {
        std::cout << "[" << i << "] = " << match[i] << " ";
    }

    std::cout << std::endl;
}

int main()
{
    using namespace std::string_literals;

    find_match("Case 0"s, R"((A+)\s*(B+)?\s*B*)"s, "AAA BBB"s);
    find_match("Case 1"s, R"((A+)\s*B*)"s, "AAA BBB"s);
    find_match("Case 2"s, R"((A+)\s*B*(B+))"s, "AAA BBB"s);
    find_match("Case 3"s, R"((A+)\s*(B+)?\s*B*(B+))"s, "AAA BBB"s);

    return 0;
}
```

With following results:

**msvc** (without fix):
```
Case 0:[1] = AAA [2] =
Case 1:[1] = AAA
Case 2:[1] = AAA [2] = B
Case 3:[1] = AAA [2] =  [3] = B
```

**msvc** (with fix), **libc++** HEAD 12.0 (Wandbox), **libstdc++** HEAD 11.0.0 (Wandbox) and browser (regex101.com) produce same results:
```
Case 0:[1] = AAA [2] = BBB
Case 1:[1] = AAA
Case 2:[1] = AAA [2] = B
Case 3:[1] = AAA [2] = BB [3] = B
```

3. I don't think that other grammars should be affected by this change as _Better_match() is used only in context when a decision whether new match should replace already found one should be made (after finishing lookup/sub-lookup during processing _N_end node).

Also, based on mentioned "The Leftmost Longest Rule" it seems that step 6 there covers an extension of capturing with additional groups, if I understand that statement correctly.

**Comment by strega-nil-ms** [comment]
The boost documentation links "leftmost" with "leftmost longest", and so I do believe that ECMAScript also follows the leftmost longest rule.

However, I believe that this is an incorrect check. Leftmost longest does not consider _how many_ groups are matched; only, from front to back, the length of each group and its location. The issue with the current check is _not_ that it doesn't check how many groups are matched - it is that it doesn't consider "non-matching" groups to be worse than matching groups.

As I was working on this, I also noticed a bug in `_Better_match()` - it considers the _rightmost_ match to be better:

```cxx
            if (_Res._Grps[_Ix]._Begin != _Tgt_state._Grps[_Ix]._Begin) {
                return _STD distance(_Begin, _Res._Grps[_Ix]._Begin)
                     < _STD distance(_Begin, _Tgt_state._Grps[_Ix]._Begin);
            }
```

_should be_

```cxx
            if (_Res._Grps[_Ix]._Begin != _Tgt_state._Grps[_Ix]._Begin) {
                return _STD distance(_Begin, _Tgt_state._Grps[_Ix]._Begin)
                     < _STD distance(_Begin, _Res._Grps[_Ix]._Begin);
            }
```

With this bug fixed, we can find a place where the code you've written fails:

regex = `/(A+)\s*(B+)?\s*B*(B*)?/`
string = `AAA BBB`

The correct match for this should be `|AAA|BBB|-|`, however, your code gives `|AAA|-|BBB|`.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
<details>
<summary><code>tr1/tests/regex2</code> is failing:</summary>

```
Standard Output:
--
Testing ECMAScript
Testing UNIX BRE
Testing grep
Testing UNIX ERE
 GOT "2 0 6 0 6" != "2 0 6 5 6"
 FAIL test 1213 at line 662 in C:\a\1\s\tests\tr1\tests\regex2\test.cpp: (aa|aabaac|ba|b|c)* |aabaac|
Testing egrep
 GOT "2 0 6 0 6" != "2 0 6 5 6"
 FAIL test 1557 at line 662 in C:\a\1\s\tests\tr1\tests\regex2\test.cpp: (aa|aabaac|ba|b|c)* |aabaac|
Testing awk
 GOT "2 0 6 0 6" != "2 0 6 5 6"
 FAIL test 1902 at line 662 in C:\a\1\s\tests\tr1\tests\regex2\test.cpp: (aa|aabaac|ba|b|c)* |aabaac|
FINISHED testing <regex>, part 2
***** 3 erroneous test cases in <regex>, part 2 *****
***** 1980 successful test cases in <regex>, part 2 *****
#FAILED: <regex>, part 2
--
```
</details>

File: stl/inc/regex
Comment (strega-nil-ms) [inline stl/inc/regex:3428]: do we want to `s/Unix/POSIX`?

File: stl/inc/regex
Comment (StephanTLavavej) [inline stl/inc/regex:3439]: I believe this logic is incorrect (although I don't immediately know what all of the problems are).

First, although it is not an authoritative reference, https://en.cppreference.com/w/cpp/regex/syntax_option_type describes POSIX as "leftmost longest". I'm surprised to see "rightmost" mentioned in this implementation. (From what I've read, ECMAScript is not "leftmost longest".) Boost explains the Perl/ECMAScript behavior at https://www.boost.org/doc/libs/1_80_0/libs/regex/doc/html/boost_regex/syntax/perl_syntax.html#boost_regex.syntax.perl_syntax.what_gets_matched and the POSIX behavior at https://www.boost.org/doc/libs/1_80_0/libs/regex/doc/html/boost_regex/syntax/leftmost_longest_rule.html .

Second, the cppreference article depicts the output:
```
Searching for .*(a|xayy) in zzxayyzz:
 ECMA (depth first search) match: zzxa
 POSIX (leftmost longest)  match: zzxayy
```

However, both `main` and this PR print:
```
Searching for .*(a|xayy) in zzxayyzz:
 ECMA (depth first search) match: zzxayy
 POSIX (leftmost longest)  match: zzxayy
```

In this example with this PR, the final `if`-statement is saying "they start at the same point, so prefer the longer one".

(As an aside, this PR's logic is structured and commented in a clear, comprehensible way, which is great! My only concern is matching what the grammars specify.)

File: tests/std/include/test_regex_support.hpp
Comment (StephanTLavavej) [inline tests/std/include/test_regex_support.hpp:120]: Need to include `<vector>`.

File: tests/std/include/test_regex_support.hpp
Comment (StephanTLavavej) [inline tests/std/include/test_regex_support.hpp:135]: (Occurs below) These messages are saying `should_capture`, but this function is `should_capture_groups` and takes additional arguments. If we fail, I think all of them should print out the groups, and also the value of `flags` as an `int`.

File: tests/std/include/test_regex_support.hpp
Comment (StephanTLavavej) [inline tests/std/include/test_regex_support.hpp:143]: `size_t` is unsigned, so this should say `%zu` (as `should_capture()` does). Occurs below.

File: tests/std/include/test_regex_support.hpp
Comment (StephanTLavavej) [inline tests/std/include/test_regex_support.hpp:163]: Suggestion: This seems confusing, because it doesn't look like the callsites. `|` means "alternations" for regexes, but here you're using it with an unofficial meaning to separate capture groups.

Would formatting the groups as `{"group1","group2","group3"}` be better?

File: tests/std/tests/VSO_0000000_regex_use/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/VSO_0000000_regex_use/test.cpp:550]: This test is already `using namespace std;`. Occurs below.

Comment (StephanTLavavej) [inline tests/std/tests/VSO_0000000_regex_use/test.cpp:550]: Should probably include `<vector>` even if `<test_regex_support.hpp>` already does so.

File: tests/std/tests/VSO_0000000_regex_use/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/VSO_0000000_regex_use/test.cpp:549]: This should be followed by a comment mentioning the bug's title (like `test_gh_993()` immediately below).

File: tests/std/tests/VSO_0000000_regex_use/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/VSO_0000000_regex_use/test.cpp:557]: I think we need additional test coverage to exercise `_Better_match`, especially scenarios where ECMAScript and POSIX Extended accept the same syntax but have different ideas about what should be captured. I think we should also manually verify agreement with Boost/libstdc++/libc++, or understand any remaining divergence. (e.g. I observed MSVC, libstdc++, and libc++ getting 3 sets of answers for the subject `"WXconcatenateYZ"` with the pattern `".*(cat|concatenate)"` https://godbolt.org/z/bza5nrYcM , woof :dog:).

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/regex
Comment (strega-nil-ms) [reply stl/inc/regex:3439]: hmm, yeah, I apparently misunderstood something as I was cleaning up the logic; I think the existing logic was worse, but yeah; not sure what the best way forward is.

**Comment by StephanTLavavej** [comment]
We talked about this at the weekly maintainer meeting. Although this PR in its current state does fix the original scenario (thanks @strega-nil-ms!), there are remaining scenarios that are unfixed, and we're uncertain about the full extent of the problem. We believe it's highly risky to merge changes that alter runtime behavior (in an old and widely-used component) without *fully* understanding the problems and having high confidence that the fix is both correct and complete. We'll keep the original bug #731 open, since this is a conformance bug that we'd like to fix someday (probably through a complete overhaul in the vNext timeframe), but at the moment we can't merge this PR as-is.

**Pull request closed without merging**
